### PR TITLE
Add InferNotNullThroughJoin rule

### PR DESCRIPTION
**Pull request by fornaix**
This rule is to add filter by join type and join key, which could reduce useless data before join or exchange.

### CODE REVIEW

**Comment by Praveen2112** [comment]
Hi !! Thanks for raising this PR.. There was also a similar PR #2415 which was trying to capture the same patter but we faced regression for certain queries bcoz of redundant filters which were added because of this patch. Benchmark is also added as a comment

**Comment by fornaix** [comment]
> Hi !! Thanks for raising this PR.. There was also a similar PR #2415 which was trying to capture the same patter but we faced regression for certain queries bcoz of redundant filters which were added because of this patch. Benchmark is also added as a comment

Thank you for telling me that there was a similar PR. I sent this PR because skewed join happended in some case. So I try to resolve it.

e.g. `SELECT a.id, a.col1, b.col2 FROM a,b WHERE a.id=b.id`. When table a or b has a lot of null column `id` records, they will all exchange into one node and cause skew.

This PR which is aimed to resolve this problem does not modify the existing PredicatePushDown rule but come up with a new rule. Cooperating with PredicatePushDown rule, this new rule could push join-key-not-null predicate related to join type into FilterNode above TableScanNode and avoid skew.

Could you please review this PR and see which solution is more suitable. Thanks!

**Comment by Praveen2112** [comment]
Actually the rule based approach looks a lot cleaner than the previous implementation but I'm afraid we might see the same regression here. For certain benchmark queries which doesn't have a null value in the join columns, we might be applying a redundant filter causes the regression.

**Comment by sopel39** [comment]
> Could you please review this PR and see which solution is more suitable. Thanks!

I think we need to make sure `is not null` checks do not bring a lot of overhead when there are no nulls in data, see: https://github.com/prestosql/presto/pull/2415#issuecomment-575605261

**Comment by fornaix** [comment]
> > Could you please review this PR and see which solution is more suitable. Thanks!
> 
> I think we need to make sure `is not null` checks do not bring a lot of overhead when there are no nulls in data, see: [#2415 (comment)](https://github.com/prestosql/presto/pull/2415#issuecomment-575605261)

So can we say that `is not null` is an expensive operator and need to be optimized?
Nevertheless, I still insist that this type of rule should be applied to reduce network cost and avoid skew. Some other query engines already have the similar rule (e.g. `InferFiltersFromConstraints` in Spark)

**Comment by raunaqmorarka** [comment]
@sopel39 @Praveen2112 could we consider bringing this change or https://github.com/prestosql/presto/pull/2415 behind a flag until the overheads of not-null checks are reduced ?
I see that prestodb added this behind a flag recently https://github.com/prestodb/presto/pull/14578

**Comment by sopel39** [comment]
> @sopel39 @Praveen2112 could we consider bringing this change or #2415 behind a flag until the overheads of not-null checks are reduced ?

I'm not sure users will be able to determine when the flag should be turned on (it's very technical flag)

**Comment by fornaix** [comment]
@sopel39  @raunaqmorarka  Two configs are added to avoid the overheads when there are no nulls in data. One is to enable this optimize rule (default false). Another one is to control whether should insert not null check for a key. Only if the nulls fraction greater or equal than the threshold (default 50%), not null predicate would be inserted.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/InferNotNullThroughJoin.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    private List<Symbol> joinKeys(JoinNode joinNode, Function<EquiJoinClause, Symbol> extractor, Set<Symbol> filteredSymbols)
+    {
+        return joinNode.getCriteria().stream()
+                .map(extractor)
+                .filter(key -> !filteredSymbols.contains(key))
+                .collect(Collectors.toList());
+    }
+
+    private Expression inferPredicatesIfNecessary(Expression filter, List<Symbol> symbols, PlanNodeStatsEstimate statsEstimate, double threshold)
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/InferNotNullThroughJoin.java:105]: that's a good idea. We need to re-benchmark it

**Comment by findepi** [comment]
> One is to enable this optimize rule (default false). Another one is to control whether should insert not null check for a key. Only if the nulls fraction is greater than or equal to the threshold (default 50%)

With stats-based approach, can we enable this by default?

**Comment by fornaix** [comment]
> > One is to enable this optimize rule (default false). Another one is to control whether should insert not null check for a key. Only if the nulls fraction is greater than or equal to the threshold (default 50%)
> 
> With stats-based approach, can we enable this by default?

I'm not sure whether we should enable this rule by default. This rule does bring overheads for the not-null check. It might be better to understand the trade-off between benefits and overheads before enabling this rule. So it is off by default.
Could you give some detailed reasons for enabling this rule by default?

File: presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java [hunk 2]
```diff
@@ -616,7 +617,13 @@ public PlanOptimizers(
                         ruleStats,
                         statsCalculator,
                         estimatedExchangesCostCalculator,
-                        ImmutableSet.of(new ReorderJoins(metadata, costComparator))));
+                        ImmutableSet.of(new ReorderJoins(metadata, costComparator))),
+
+                new IterativeOptimizer(
```
[the remark this answers is not in the document]
Comment (fornaix) [reply presto-main/src/main/java/io/prestosql/sql/planner/PlanOptimizers.java]: @martint Thanks, that was helpful. Updated

**Comment by findepi** [comment]
> Could you give some detailed reasons for enabling this rule by default?

Sure.
We strive to provide best behavior possible -- out of the box.
When an optimization is enabled by default, it is going to benefit all users.
When it's off by default, only those few users will enjoy it, who actually learn about the toggle and take time to enable it.

> I'm not sure whether we should enable this rule by default. This rule does bring overheads for the not-null check. 

I am aware of the incurred overhead. That's the reason why it's being stats-based.

> It might be better to understand the trade-off between benefits and overheads before enabling this rule. 

Are you proposing to have it on after some time?
That's might be OK too, as long as we make sure we do not forget to actually make a switch.
Also, what would be the conditions to make the switch?

**Comment by fornaix** [comment]
> We strive to provide best behavior possible -- out of the box.
When an optimization is enabled by default, it is going to benefit all users.
When it's off by default, only those few users will enjoy it, who actually learn about the toggle and take time to enable it.

@findepi  Good views! I tend to turn it on by default now. 
If there are no other opinions, I will enable this rule by default.

**Comment by findepi** [comment]
> I will enable this rule by default.

thanks @fornaix !

> If there are no other opinions

@martint @sopel39

**Comment by sopel39** [comment]
Thanks @fornaix. 

This PR needs benchmarking before we merge it.

**Comment by fornaix** [comment]
[Benchmark_InferNull.pdf](https://github.com/prestosql/presto/files/5078647/Benchmark_InferNull.pdf)

@martint @findepi  @sopel39  Here are the benchmark results on 10x nulls basis points data(The original nulls fraction is too low to exceed the default threshold).

It doesn't seem to improve performance much. But it will help to reduce skew.

Here is a simple case
```sql
WITH tb AS (
		SELECT ss_customer_sk, ss_sold_date_sk
		FROM store_sales
		WHERE ss_sold_date_sk IS NULL
			OR ss_sold_date_sk % 2 = 0
	)
SELECT 1.0 * count_if(ss_customer_sk IS NULL) / COUNT(1)
	, 1.0 * count_if(ss_sold_date_sk IS NULL) / COUNT(1)
FROM tb;
/*
nulls fraction:

_col0 | _col1
-------+-------
 0.5   | 0.6
*/

set session optimize_nulls_in_join=false;
WITH tb AS (
		SELECT ss_customer_sk, ss_sold_date_sk
		FROM store_sales
		WHERE ss_sold_date_sk IS NULL
			OR ss_sold_date_sk % 2 = 0
	)
SELECT COUNT(1) AS cnt, ss_customer_sk
FROM tb
	JOIN tb USING (ss_customer_sk, ss_sold_date_sk)
GROUP BY ss_customer_sk
ORDER BY cnt DESC
LIMIT 10;
/*
failed: Query exceeded per-node user memory limit of 20GB...
*/

set session optimize_nulls_in_join=true;
-- same query
/*
  cnt  | ss_customer_sk
-------+----------------
 15138 |         272024
 14987 |         125502
 14869 |         293849
 14759 |          91975
 14736 |          62381
 14729 |         171809
 14728 |         116087
 14675 |         311074
 14579 |          93385
 14572 |         242158
(10 rows)

Query 20200815_140555_16330_czfa4, FINISHED, 112 nodes
Splits: 9,320 total, 9,320 done (100.00%)
0:08 [4.28B rows, 1.29GB] [532M rows/s, 164MB/s]
*/
```

**Comment by sopel39** [comment]
Could you rebase? I will also run our benchmark suite

**Comment by fornaix** [comment]
> Could you rebase? I will also run our benchmark suite

Done.

**Comment by sopel39** [comment]
IMO we should close this given lack of good improvements (see also https://github.com/trinodb/trino/pull/2415)

**Comment by findepi** [comment]
@sopel39 if so, we should also close https://github.com/trinodb/trino/issues/2392

**Pull request closed without merging**